packages/odyssey-react-mui/src/labs/FileUpload.tsx

```diff
@@ -0,0 +1,294 @@
+  border: `1px dashed ${odysseyDesignTokens.HueNeutral300}`,
+  borderRadius: odysseyDesignTokens.BorderRadiusMain,
+  transition: `border-color ${odysseyDesignTokens.TransitionTimingMain}, box-shadow ${odysseyDesignTokens.TransitionTimingMain}`,
+
+  "&:hover": {
+    borderColor: odysseyDesignTokens.HueNeutral700,
+  },
+
+  "&:has(input:focus)": {
+    borderStyle: "solid",
+    borderColor: odysseyDesignTokens.FocusOutlineColorPrimary,
+    boxShadow: `0 0 0 1px ${odysseyDesignTokens.FocusOutlineColorPrimary}`,
+    outline: `${odysseyDesignTokens.FocusOutlineWidthMain} ${odysseyDesignTokens.FocusOutlineStyle} transparent`,
+    outlineOffset: odysseyDesignTokens.FocusOutlineOffsetTight,
+  },
+
+  "&:has(input:disabled)": {
+    backgroundColor: odysseyDesignTokens.HueNeutral50,
+    border: `1px solid ${odysseyDesignTokens.BorderColorDisabled}`,
+    color: odysseyDesignTokens.TypographyColorDisabled,
+
+    "&:hover": {
+      borderColor: odysseyDesignTokens.BorderColorDisabled,
+    },
+  },
+}));
+
+const ButtonAndInfoContainer = styled.div({
+  display: "flex",
+  flexDirection: "column",
+  alignItems: "center",
+  justifyContent: "center",
+});
+
+export type FileUploadProps = {
+  /**
+   * an array of file types the user is able to upload. @see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept#unique_file_type_specifiers for examples
+   */
+  acceptedFileTypes?: string[];
+  /**
+   * The label for the `input` element.
+   */
+  label: string;
+  /**
+   * Function that is called when the list of ifles to upload is changed
+   */
+  onChange: (files: File[]) => void;
+  /**
+   * Either `single` or `multiple`. If `multiple`, multiple files can be uploaded
+   */
+  type?: (typeof fileUploadTypes)[number];
+  /**
+   * Either `button` or ``. If `multiple`, multiple files can be uploaded
+   */
+  variant: (typeof fileUploadVariants)[number];
+} & Pick<
+  FieldComponentProps,
+  | "errorMessage"
+  | "hint"
+  | "HintLinkComponent"
+  | "id"
+  | "isDisabled"
+  | "isOptional"
+>;
+
+type RemoveFileFromFilesToUploadList = (name: string) => void;
+
+const FileUpload = ({
+  acceptedFileTypes,
+  errorMessage,
+  id,
+  isDisabled = false,
+  isOptional,
+  hint,
+  HintLinkComponent,
+  label,
+  onChange,
+  type,
+  variant,
+}: FileUploadProps) => {
+  const odysseyDesignTokens = useOdysseyDesignTokens();
+  const { t } = useTranslation();
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [filesToUpload, setFilesToUpload] = useState<File[]>([]);
+
+  useEffect(() => {
+    onChange(filesToUpload);
+  }, [filesToUpload, onChange]);
+
+  const updateFilesToUpload = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      const { files } = event.target;
+
+      if (files && files.length > 0) {
+        const mergedFiles =
+          type === "multiple" ? [...filesToUpload, ...files] : [...files];
+
+        setFilesToUpload(mergedFiles);
```

Comment:
Do we need `[...files]`?
```suggestion
        const mergedFiles =
          type === "multiple" ? [...filesToUpload, ...files] : files;

        setFilesToUpload(mergedFiles);
```

Also, what about this?
```suggestion
        const mergedFiles =
          type === "multiple" ? files.concat(filesToUpload) : files;

        setFilesToUpload(mergedFiles);
```

Reply:
@KevinGhadyani-Okta Yeah, the type for `files` here is `FileList` but, it needs to be coerced into an array of `File`s

Reply:
We should probably add a comment then. It wasn't clear looking at the code, and someone is sure to change this in the future.

Reply:
Maybe better if we did `files) satisfies File[] as File[]`.